### ISSUE TITLE
benchmark times

### ISSUE THREAD

**Issue by dustinvtran** [airoldilab/sgd#38]
Just get some rough heuristics for how it performs now

**Comment by lantian2012** [comment]
`sgd` performs worse with high dimensional dataset. 
```
N <- 1e4
d <- 1e3
Unit: seconds
 expr      min       lq     mean   median       uq      max neval
  sgd 65.71423 71.59921 74.06408 77.48419 78.23900 78.99381     3
  glm 18.75750 19.19480 19.85834 19.63210 20.40876 21.18541     3
```

```
N <- 1e4
d <- 5
Unit: milliseconds
 expr      min       lq     mean   median       uq      max neval
  sgd 26.19169 27.29816 28.82894 29.35097 30.15182 31.15207     5
  glm 28.25475 29.27066 30.59445 29.79457 29.99383 35.65846     5
```

If `glm` uses least squares, it should have Nd^2 complexity. Why should `implicit` scales worse?

**Comment by lantian2012** [comment]
I'm looking into the performance of sgd before we put it into the package.

**Comment by lantian2012** [comment]
The code to run the sgd without the package is somehow broken. I find at working version at 1472b6d011219abde3601c709f2600eecf0315a5.  At that time `implicit` was much faster. Here's benchmark result:

```
normal.e = normal.experiment(niters=1e5, p=100)
microbenchmark(run.test.normal.imp.c(normal.e, normal.data), glm(normal.data$Y~normal.data$X-1), times=4)

 expr       min        lq      mean    median
 run.test.normal.imp.c(normal.e, normal.data)  396.1928  405.2936  421.1427  417.4138
       glm(normal.data$Y ~ normal.data$X - 1) 3731.6620 3773.9635 3821.0964 3839.1495
        uq       max neval
  436.9919  453.5506     4
 3868.2292 3874.4244     4
```

The performance of `implicit` package is similar to `sgd` package.

**Comment by dustinvtran** [comment]
One thing is that here the learning rate is always a diagonal matrix (even in the `one-dim` and `one-dim-eigen` cases). This was done in order to make the output of the learning rate classes consistent; I'll see if I can extend it so that the one-dimensional cases truly output a scalar

**Comment by lantian2012** [comment]
One workaround is to define our own `operation*`. We can check the learning rate type and dispatch to a corresponding multiplication method.

**Comment by dustinvtran** [comment]
Cool idea! I'll add it to the learning rate class.

**Comment by dustinvtran** [comment]
Err this is more complicated than I thought. It won't work unless we introduce yet another class with derived classes from a base class.`at` is of type `mat` right now, and we shouldn't overload the default `operator*()` for two `mat` types.

**Comment by lantian2012** [comment]
I finally got profiling for c code to work. I'll run some experiments tonight to check the performance. I don't think we need to make the structure too complicated if the learning rate is not the performance bottleneck.

**Comment by lantian2012** [comment]
According to profiling, as we construct a new d*d learning rate matrix at each iteration, constructing the matrix takes lots of time.
However, after I broke the code and forced the learning rate to be a scaler, the complexity is still nd^2. There should be another bottleneck.

**Comment by lantian2012** [comment]
I have identified the problem.

1. Allocating learning rate matrix is taking the majority of computation time (over 70%). I think we should add a new class to store the return value for `learning_rate()`.

2. For `one-dim` learning rate, initilization calculates the eigen values of the design matrix. I think we should come up with another strategy. 

3. The current implementation checks the rank of the design matrix, which also takes a lot of time.

**Comment by ptoulis** [comment]
sounds good.

i am relieved that it looks easily flexible.

**Comment by lantian2012** [comment]
I think only glm will need to check the rank of design matrix. I plan to add rank checking as a flag for glm model. 
I'm working on the fix for the data type of learning rate. I benchmarked the performance. 

```
N <- 1e4
d <- 1e3
Unit: seconds
 expr       min        lq     mean    median        uq       max neval
  sgd  2.426677  2.506226  2.55444  2.511836  2.554792  2.772668     5
  glm 18.546360 18.562030 18.73236 18.728130 18.793832 19.031425     5
```

**Comment by dustinvtran** [comment]
Excellent news. :) I went ahead and also stored Idiag as a vector instead of a diagonal matrix. There's one more clear optimization, which is to reduce storage/calculation of the diagonal learning rate as a vector instead of diagonal matrix too. I tried but couldn't get it to work, c.f., https://github.com/airoldilab/sgd/blob/26eba14ad984903b1ae09a12275bba6d8b1546ae/src/learningrate.h#L16-L52

**Comment by ptoulis** [comment]
now we are in business..:)

**Comment by dustinvtran** [comment]
Will push benchmark source file soon. :)
```{R}
library(sgd)
library(glmnet)
library(microbenchmark)

# Dimensions
N <- 1e5
d <- 1e2

# Generate data.
set.seed(42)
X <- matrix(rnorm(N*d), ncol=d)
theta <- rep(5, d+1)
eps <- rnorm(N)
y <- cbind(1, X) %*% theta + eps
dat <- data.frame(y=y, x=X)

# Benchmark!
benchmark <- microbenchmark(
  sgd=sgd(y ~ ., data=dat, model="lm",
          sgd.control=list(method="implicit")),
  lm=lm(y ~ ., data=dat),
  glmnet=glmnet(X, y, alpha=1, standardize=FALSE, type.gaussian="naive"),
  times=10L
)
benchmark
## Output (for 2.6 GHz, Intel Core i5)
## Unit: seconds
##    expr      min       lq     mean   median       uq      max neval
##     sgd 1.302610 1.393843 1.456990 1.459721 1.483716 1.709765    10
##      lm 1.650819 1.706835 1.776254 1.763900 1.786094 1.938011    10
##  glmnet 2.451650 2.552243 2.617811 2.623174 2.635034 2.863428    10
```
